packages/pagination/src/index.ts

```diff
@@ -0,0 +1,216 @@
+      }
+      return props;
+    });
+    return [paginationProps, page, setPage as Setter<number>];
+  } else {
+    const onKeyUp = (pageNo: number, ev: KeyboardEvent) =>
+      ((
+        {
+          ArrowLeft: () => setPage(p => p - 1),
+          ArrowRight: () => setPage(p => p + 1),
+          Home: () => setPage(1),
+          End: () => setPage(opts().pages),
+          Space: () => setPage(pageNo),
+          Return: () => setPage(pageNo)
+        }[ev.key] || noop
+      )());
+    // create all pages so we can reuse the objects
+    const pages: PaginationProps = [...Array(opts().pages)].map((_, i) =>
+      ((pageNo: number) =>
+        Object.defineProperties(
+          {
+            children: pageNo.toString(),
+            onClick: [setPage, pageNo] as const,
+            onKeyUp: [onKeyUp, pageNo] as const
+          },
+          {
+            "aria-current": {
+              get: () => (page() === pageNo ? "true" : undefined),
+              set: noop,
+              enumerable: true
+            },
+            "page": {
+              value: pageNo,
+              enumerable: false,
+            },
+          }
+        ))(i + 1)
+    );
+    const first = Object.defineProperties({
+      onClick: [setPage, 1] as const,
+      onKeyUp: [onKeyUp, 1] as const,
+    } as unknown as PaginationProps[number], {
+      disabled: { get: () => page() <= 1, set: noop, enumerable: true },
+      children: { get: () => opts().firstContent, set: noop, enumerable: true },
+      page: { value: 1, enumerable: false },
+    });
+    const back = Object.defineProperties({
+      onClick: () => setPage(p => (p > 1 ? p - 1 : p)),
+      onKeyUp: (ev: KeyboardEvent) => onKeyUp(page() - 1, ev)
+    } as unknown as PaginationProps[number], {
+      disabled: { get: () => page() <= 1, set: noop, enumerable: true },
+      children: { get: () => opts().prevContent, set: noop, enumerable: true },
+      page: { get: () => Math.min(1, page() - 1), enumerable: false },
+    });
+    const next = Object.defineProperties({
+      onClick: () => setPage(p => (p < opts().pages ? p + 1 : p)),
+      onKeyUp: (ev: KeyboardEvent) => onKeyUp(page() - 1, ev)
+    } as unknown as PaginationProps[number], {
+      disabled: { get: () => page() >= opts().pages, set: noop, enumerable: true },
+      children: { get: () => opts().nextContent, set: noop, enumerable: true },
+      page: { get: () => Math.max(opts().pages, page() + 1), enumerable: false },
+    });
+    const last = Object.defineProperties({
+      onClick: () => setPage(opts().pages),
+      onKeyUp: (ev: KeyboardEvent) => onKeyUp(opts().pages, ev)
+    } as unknown as PaginationProps[number], {
+      disabled: { get: () => page() >= opts().pages, set: noop, enumerable: true },
+      children: { get: () => opts().lastContent, set: noop, enumerable: true },
+      page: { get: () => opts().pages, enumerable: false },
+    });
+    const paginationProps = createMemo<PaginationProps>(() => {
```

Comment:
hmm
having the same code in two places doesn't sit well with me, but it doesn't affect runtime so that's not a blocker
maybe it's something to look into after the merge if one has time